### ISSUE TITLE
Add z-levels for roads & railways

### ISSUE THREAD

**Issue by hixi** [geometalab/osmaxx-frontend#420]

**Comment by das-g** [comment]
Add a column to the `road_l` and `railway_l` reflecting the value of [OSM tag `layer=*`]
(http://wiki.openstreetmap.org/wiki/Key:layer).

OSM wiki says:
> All ways without an explicit value are assumed to have layer 0.

But I guess ArcGIS and QGIS work better with explicit instead of implicit values. (@sfkeller: Is that assumption correct?) Thus, for OSM features where the tag is absent, use value `0` in the new column.

**Comment by sfkeller** [comment]
I suggest to use value NULL (if it's integer, else '' if it's TEXT) in the new column if OSM features' tag is absent in order to distinguish it to 0. Remind that there exist also tags "layer=0".

**Comment by das-g** [comment]
> [Remember] that there exist also tags "layer=0".

They [do exist](http://taginfo.openstreetmap.org/tags/layer=0), even though [discouraged](http://wiki.openstreetmap.org/wiki/Key:layer):
> Values between -5 to 5 are used, **0 should not be used explicitly.**

(emphasis mine)

But if I understand
> All ways without an explicit value are assumed to have layer 0.

on the wiki correctly, there is no semantic difference between the `layer=*` tag being absent and `layer=0`. So representing them the same way in our export seems unproblematic and even preferable.

My question is what is best fit for getting ArcGIS and QGIS to draw ways on top of each other correctly. I guess that `NULL` isn't.

**Comment by sfkeller** [comment]
The modeling heavily depends if it's an INT or TEXT.

**Comment by das-g** [comment]
More than 99.86%  of [`layer=*` values in use](http://taginfo.openstreetmap.org/keys/layer#values) can be interpreted as integers. So an INT column makes sense, IMHO. (Be sure to map strings not parseable to `0`.)

**Comment by das-g** [comment]
All 4 tables produced by the osm2pgsql import (`osm_line`, `osm_point`, `osm_polygon` and `osm_roads`) have both, a `layer` column **and** a `z_order` column.

The `layer` column (of SQL type TEXT, if [our test metadata](https://github.com/geometalab/osmaxx-frontend/blob/7e3c65a4c48be3096427f847bec9fe3c4cd6306d/tests/inside_worker_test/declarative_schema/osm_models.py#L49) is accurate) is most probably **the verbatim value of OSM's `layer=*` tag if present**, otherwise `NULL`. It should be (and usually is) a string representing a nonzero integer between -5 and 5. Technically, it can be any text, and a small fraction of tags (< 0.14%) actually is some non-integer strings.

The `z_order` column (of SQL type SMALLINT [according to our test metadata](https://github.com/geometalab/osmaxx-frontend/blob/7e3c65a4c48be3096427f847bec9fe3c4cd6306d/tests/inside_worker_test/declarative_schema/osm_models.py#L99)) takes the `layer` value (if present and interpretable as number (integer or float)) into account, but admixes whether the object has been tagged as `bridge=yes`, `tunnel=yes` and/or as one of several known `highway=*` tags with various wight. The result is an integer number, usually between -60 and 69. If an object has no tags contributing to the calculation, the value is 0.

While the `z_order` _is_ related to relative physical altitude of overlapping or close features, it also has summands in it that are related to (cartographic) "importance" rather than relative physical position. **`z_order` is clearly meant as draw order for visual display** of the data (map rendering), not as an accurate information about actual relative position of features.

@sfkeller, @cdeUG, @geoSLG:
What information should be added as additional OSMaxx layer attribute(s)?
* only `z_order`<br />or
* only `layer` converted to integer number (if possible, else `0`)<br />or
* both (as separate attributes)<br />or
* something else

Should these attributes be added to OSMaxx layers `road_l` and `railway_l`, or to others, too?

**Comment by das-g** [comment]
For those interested: The calculation logic of `z_order` is located at https://github.com/geometalab/osmaxx-frontend/blob/7e3c65a4c48be3096427f847bec9fe3c4cd6306d/osmaxx/conversion/converters/converter_gis/bootstrap/styles/style.lua#L13

An AFAICT equivalent but better commented implementation can be found at https://github.com/openstreetmap/osm2pgsql/blob/6551f935bafe638aaf42dfe8065c97dbdd3d1ef8/style.lua#L20

**Comment by hixi** [comment]
Actually, our style.lua is situated here: https://github.com/geometalab/docker-continuous-osm2pgsql/blob/develop/styles/tag_transform_style.lua

**Comment by das-g** [comment]
I just noticed that `z_order` is already now (and has been for a long time) present in our OSMaxx output layers `nonop_l`, `road_l` and `railway_l`.

I've looked at those attributes on actual OSMaxx exports and they seem to have sensible (as in, consistent with my explanation above according to their OSM tagging) values.

**Comment by das-g** [comment]
Moving back to Backlog, as I can't remember whether this request was for enabling data analysis or improving map rendering. I guess we'll have to wait until @cdeUG is back. (IIRC, she was the one who requested this at our face-to-face meeting in Zürich on 2016-03-31.)

**Comment by geoSLG** [comment]
Yes `z-order` values are good and have always been part of the data model. I found the values in the newer exports as well as in the older ones. I think we can perfectly work with the `z-order `values. At the meeting we were talking about  `z-level ` values.
@cdeUG do you have further requirements to `z-order ` values?

values

**Comment by geoSLG** [comment]
Yes  `z-order  `values are good and have always been part of the data model. I found the values in the newer exports as well as in the older ones. I think we can perfectly work with the  `z-order `values. At the meeting we were talking about  z-level  values.
@cdeUG do you have further requirements to  z-order  values?

**Comment by das-g** [comment]
@geoSLG, by "z-level", do you mean the elevation a.k.a. ground altitude above see level? If so, that's being tracked in #421.

**Comment by geoSLG** [comment]
@das-g Nope. I didn't mean elevation. You marked that correctly as "nice to have".  Most Probably `z-level` was used as another term for `z-order `.

**Comment by cdeUG** [comment]
geoSLG, @das-g, @hixi, @sfkeller   Good issue! The `z-order` value for each single object id in my present ArcGis application may be of cartographic use for visualization of overpass an underpass of line objects, but it leads to an incredibly long feature class content list as many attribute combination will be the result for each type (several sub-types; now we have only 3: ground level, 0,0 / bridges 1,0 /  tunnels 0,1). I assume for QGIS it is the same. I cannot imagine that future OSMaxx basemap user will be ready to deal with it, not to mention the symbol level numbers for all feature classes which have to be considered too (for line intersections and class level I defined symbol level numbers).

Quick look back: In the first ArcGis cartographic application of HSR (example Rapperswil) `z-order` indices were not taken into consideration for mapping (even if it was always part of the SQL-database), I don't know if there were some reflection on it behind. Right @das-g we discussed it at the meeting on March in Zurich to distinguish the meaning of `z-order` (order in space) and `z-level` (elevation). But anyway, me too I didn't reflect the issue more deeply since then. But given the above reasons (complexity of handling) I did never consider to make changes in terms of a more differentiated value field selection since then. So we should probably once discuss in greater depth, if we can live with this specific mapping disadvantage or not.

But most important for the new release, I recommend to keep the `z-order` value our OSMaxx database - as some user might like to use it for better cartographic quality in one or other case.

Whether for the project we should make of such fundamental changes for map rendering of `road_l` and `railway_l` features - an in-depth discussion is needed. Actually I cannot estimate the effort it may take and if we would ever get a satisfactory and user-friendly result.
